README.md
# kubernetes-ingress-info

This project provides a simple REST service that can be deployed to any Kubernetes cluster
that reports whether or not a `host` exists in any `Ingress` deployed to that cluster.

Its **important** that this API does **not** report on the *health* of any given `Ingress` `host`
but specifically whether or not ANY `Ingress` currently exists on the cluster where `kubernetes-ingress-info`
is deployed, regardless of if its backend is actually up. It is intended to be a very lightweight and open
endpoint to interrogate for this status.

This tool was specifically developed to serve as a endpoint for various cloud load-balancers
and DNS services to be able to auto-detect *which* Kubernetes clusters are valid for a specific `host` FQDN.
This can aid in workload migration across various clusters or just enable workloads to be deployed to subsets
of clusters out of many organizational clusters, while letting upstream DNS and or glsb based devices react
accordingly as many of these platforms rely on simple HTTP 200 OK type of "checks".

* [Use case](#usecase)
* [Install/Setup](#req)
* [Example](#example)
* [API](#api)
* [Caching](#cache)
* [Usage](#usage)


## <a name="usecase"></a>Use case:

This tool was specifically developed to serve as a endpoint for various cloud load-balancers
and DNS services to be able to auto-detect *which* Kubernetes clusters are valid for a specific `host` FQDN.

Lets say you have two apps available at **a.b.com** and **x.y.com** and you have 3 available Kubernetes
it could potentially be deployed on, *clusterA*, *clusterB* and *clusterC*. Generally upstream from your cluster *LoadBalancers*
you will have another cloud load balancer device or you are just using some sort of DNS device to control what cluster *LoadBalancer* IPs are relevant for each application at any given time. So what you can do in this case is list all possible cluster IPs as possible targets for those FQDNs hostname, but enable/disable them based on calls to `kubernetes-ingress-info` which
resides on all clusters. For this to work, you should ensure each possible target cluster `Ingress Controller` that is behind a `LoadBalancer` has `kubernetes-ingress-info` accessible via its own unique `Ingress`.

![diag](/docs/diag1.png "Diagram1")

## <a id="req"></a>Install/Setup

Helm chart:  
https://github.com/boarder981/k8s-ingress-info

Run via Docker:  
https://hub.docker.com/r/bitsofinfo/kubernetes-ingress-info

Otherwise:

**Python 3.6+**

Dependencies: See [Dockerfile](Dockerfile)

When deploying to a Kubernetes cluster the `Pod` that `kubernetes-ingress-info` is
deployed as needs to run as a `serviceAccount` with (at a minimum) a binding to a
`[Cluster]Role` with the following RBAC permissions:

```
...
rules:
- apiGroups: ["extensions"]
  resources: ["ingresses"]
  verbs: ["list"]
```



## <a name="example"></a>Example:

This example will install `kubernetes-ingress-info` onto a Kubernetes cluster.
The [example.yaml](examples/example.yaml) creates a `Namespace`, `ServiceAccount`
with RBAC bindings, a `Deployment`, `Service` and an `Ingress` bound to `host: kubernetes-ingress-info.local`.

The `ServiceAccount` has the ability to `list` all `ingresses` on a cluster. You can modify
and install to your cluster however you see fit.

```
kubectl apply -f examples/example.yaml
```

Once installed, you will need a `/etc/hosts` entry or DNS setup for `kubernetes-ingress-info.local` to
point to the `LoadBalancer` configured for whatever `IngressController` consumed the example `Ingress` in `example.yaml`

Let's hit the root to get a list of all `Ingress` "hosts" on the cluster.
```
curl http://kubernetes-ingress-info.local

[
  "kubernetes-ingress-info.local",
  ....
]
```

Validate the existence of a single `host`, will return a 200
```
curl http://kubernetes-ingress-info.local/kubernetes-ingress-info.local

{
  "info": "'kubernetes-ingress-info.local' found"
}
```

How about one that does not exist? (returns a 404)
```
curl http://kubernetes-ingress-info.local/non-existant-host.com

{
  "error": "'non-existant-host.com' 404 not found"
}
```

Cleanup:
```
kubectl delete -f examples/example.yaml
```


## <a name="api"></a>API:

The API is as follows:

GET all Ingress Hosts on the cluster:
```
curl -X GET http[s]://[kubernetes-ingress-info]
```

GET a specific `host`. Returns 200 if `host` exists, or a 404 if non-existant
```
curl -X GET http[s]://[kubernetes-ingress-info]/[host]
```

Note that if you startup `info.py` with the `--host-match-on-header <headername>` value, the named HTTP request header
will be checked for a `host`. This `host` name will then be used if the `/<path>` host is not found

GET a specific `host` (via custom header). Returns 200 if `host` exists, or a 404 if non-existant
```
curl -X GET http[s]://[kubernetes-ingress-info]/anything -H '<--host-match-on-header header name>: [host]'
```

## <a name="cache"></a>Caching:

If you enable caching w/ `--enable-cache` the internal Ingress database that is constructed
after fetching all available `Ingress` from Kubernetes (and applying all specified selectors)
is cached to disk. The cache structure is a single key that stores the entire dictionary to
local disk. If running in a container this should have *minimal* storage requirements and
*does not necessitate a dedicated volume*.

We currently use the https://github.com/grantjenks/python-diskcache library due to its `get` speed
and active project activity.

## <a name="usage"></a>Usage:

```
bash-4.4$ python3 info.py -h

usage: info.py [-h] [-r LOAD_CONFIG_MODE] [-i INCLUDE_LABEL_SELECTORS]
               [-x EXCLUDE_LABEL_SELECTORS] [-n NAMESPACES]
               [-m HOST_MATCH_ON_HEADER] [-p LISTEN_PORT] [-c]
               [-e CACHE_TTL_SECONDS] [-d CACHE_DIR] [-l LOG_LEVEL]
               [-b LOG_FILE]

optional arguments:
  -h, --help            show this help message and exit
  -r LOAD_CONFIG_MODE, --load-config-mode LOAD_CONFIG_MODE
                        How the target k8s cluster config will be loaded,
                        'local' will leverage the current kubectl context from
                        '~/.kube/config', while 'cluster' will talk direct to
                        the cluster the process is executing in using the
                        Pod's configured serviceAccount which needs read
                        access to all objects of type Ingress. Default 'local'
  -i INCLUDE_LABEL_SELECTORS, --include-label-selectors INCLUDE_LABEL_SELECTORS
                        Optional comma delimited of Ingress
                        label1=value,label2=value pairs that will be used to
                        build the database of IngressInfo objects available to
                        be fetched. If specified, ONLY Ingress objects having
                        ALL specified labels will be retrieved. If not
                        specified, ALL available Ingress objects will be
                        retrieved from k8s. Excludes take precendence over
                        includes.
  -x EXCLUDE_LABEL_SELECTORS, --exclude-label-selectors EXCLUDE_LABEL_SELECTORS
                        Optional comma delimited of Ingress
                        label1=value,label2=value pairs that will be used to
                        restrict the database of IngressInfo objects available
                        to be fetched. If specified, ONLY Ingress objects NOT
                        having ANY of the specified labels will be retrieved.
                        If not specified, ALL available Ingress objects will
                        be retrieved from k8s. Excludes take precendence over
                        includes.
  -n NAMESPACES, --namespace NAMESPACES
                        Optional comma delimited of Namespaces to scope
                        Ingress fetch within
  -m HOST_MATCH_ON_HEADER, --host-match-on-header HOST_MATCH_ON_HEADER
                        Optional name of HTTP Request header who's value will
                        be inspected for an Ingress host match, should the
                        default path based /[host] not find a match.
  -p LISTEN_PORT, --listen-port LISTEN_PORT
                        Port to listen on, default 8081
  -c, --enable-cache    Optional, enabling caching (uses
                        https://github.com/grantjenks/python-diskcache)
  -e CACHE_TTL_SECONDS, --cache-ttl-seconds CACHE_TTL_SECONDS
                        Optional, cache TTL in seconds
  -d CACHE_DIR, --cache-dir CACHE_DIR
                        Optional, cache dir, default /opt/kubernetes-ingress-
                        info/cache
  -l LOG_LEVEL, --log-level LOG_LEVEL
                        log level, default DEBUG
  -b LOG_FILE, --log-file LOG_FILE
                        Path to log file, default None, STDOUT
```                        


CHANGELOG.md
# 1.0.6
* Fix bug when `--host-match-on-header` is present for `str' object has no attribute 'getHeader`

# 1.0.5
* Fix bug when `--host-match-on-header` is present

# 1.0.4
* Add new option to check for custom header to indicate the ingress host to check, if path does not match

# 1.0.3
* dummy tag

# 1.0.2
* Fixed int handling of `TTL`

# 1.0.1
* Added caching support w/ https://github.com/grantjenks/python-diskcache

# 1.0.0
* Initial version


info.py
#!/usr/bin/env python

__author__ = "bitsofinfo"

from kubernetes import client, config
import pprint
import argparse
import sys
import logging
import json
from twisted.web import server, resource
from twisted.internet import reactor, endpoints
import time
import threading
from diskcache import Cache


class IngressInfo(resource.Resource):
    isLeaf = True
    load_config_mode = "local"
    v1Api = None
    exclude_label_selectors = None
    include_label_selectors = None
    namespaces = None
    cache = None
    cache_ttl_seconds = 300
    host_match_on_header = None

    # initialize our client for proper k8s config
    def init(self):
        if self.load_config_mode == "local":
            config.load_kube_config()
        else:
            config.load_incluster_config()

        self.v1Api = client.ExtensionsV1beta1Api()

    # fetches the Ingress database
    def getIngressDb(self):

        if self.cache is not None:
            cachedIngressDb = self.cache.get(b'INGRESS_DB')
            if cachedIngressDb is not None:
                return cachedIngressDb


        fetchedIngresses = []

        # fetch across ALL namespaces
        if self.namespaces is None:
            if self.include_label_selectors is not None:
                logging.debug("getIngressDb() from k8s: fetching (namespace=*) only Ingress w/ label selectors: %s" % self.include_label_selectors)
                k8sIngresses = self.v1Api.list_ingress_for_all_namespaces(watch=False, \
                    label_selector=self.include_label_selectors)
                fetchedIngresses.extend(k8sIngresses.items)
            else:
                logging.debug("getIngressDb() from k8s: fetching (namespace=*)")
                k8sIngresses = self.v1Api.list_ingress_for_all_namespaces(watch=False)
                fetchedIngresses.extend(k8sIngresses.items)

        # fetch within namespace
        else:
            for namespace in self.namespaces:
                if self.include_label_selectors is not None:
                    logging.debug("getIngressDb() from k8s: fetching (namespace=%s) only Ingress w/ label selectors: %s" % namespace,self.include_label_selectors)
                    nsIngresses = self.v1Api.list_namespaced_ingress(namespace, watch=False, \
                        label_selector=self.include_label_selectors)
                    fetchedIngresses.extend(nsIngresses.items)
                else:
                    logging.debug("getIngressDb() from k8s: fetching (namespace=%s)" % namespace)
                    nsIngresses = self.v1Api.list_namespaced_ingress(namespace, watch=False)
                    fetchedIngresses.extend(nsIngresses.items)



        ingressDb = { 'unique_hosts':set() }
        for i in fetchedIngresses:

            excludeIngress = False
            if self.exclude_label_selectors is not None:
                for exlbl,exlblval in self.exclude_label_selectors.items():
                    ingressLabelNames = i.metadata.labels.keys()
                    if exlbl in ingressLabelNames and i.metadata.labels[exlbl] == exlblval:
                        logging.debug("getIngressDb() excluding Ingress: \"%s:%s\" due to exclude label: \"%s=%s\"" % (i.metadata.namespace,i.metadata.name,exlbl,exlblval))
                        excludeIngress = True
                        break;

            if excludeIngress:
                continue;

            for r in i.spec.rules:
                ingressInfo = { 'host':r.host.lower() }
                ingressDb['unique_hosts'].add(r.host.lower())

        if self.cache is not None:
            self.cache.set(b'INGRESS_DB', value=ingressDb, expire=self.cache_ttl_seconds)

        return ingressDb


    # The main servicing method
    def render_GET(self, request):

        toReturn = {}
        try:
            ingressDb = self.getIngressDb()

            requestUri = request.uri.decode('utf-8')

            # if just a root request, send the contents of the ingressDb
            if requestUri == "/":
                toReturn =  list(ingressDb['unique_hosts']) # convert to List for json...

            # otherwise we assume the request is checking a fqdn/host existence in ingressDb
            else:
                # assume the uri path is a host
                hostToCheck = requestUri.replace("/","")
                if hostToCheck.lower() in ingressDb['unique_hosts']:
                    toReturn = {'info':("'%s' found" % hostToCheck)}

                # if host-match-on-header was specified... check for it
                elif self.host_match_on_header and \
                     self.host_match_on_header.strip() != "" and \
                     request.getHeader(self.host_match_on_header) and \
                     request.getHeader(self.host_match_on_header).strip() != "" and \
                     request.getHeader(self.host_match_on_header).lower() in ingressDb['unique_hosts']:
                    
                    toReturn = {'info':("'%s' found" % request.getHeader(self.host_match_on_header))}

                # otherwise no match
                else:
                    toReturn = {'error':("'%s' 404 not found" % hostToCheck)}
                    request.setResponseCode(404)

        except Exception as e:
            error = "failure: " + str(sys.exc_info())
            toReturn = {'error':("500 failure: %s" % error)}
            request.setResponseCode(500)
            logging.exception("render_GET unexpected error: " + error)

        finally:
            return json.dumps(toReturn, indent=2).encode("UTF-8")


def init(load_config_mode, \
         listen_port, \
         enable_cache, \
         cache_dir, \
         cache_ttl_seconds, \
         include_label_selectors, \
         exclude_label_selectors, \
         namespaces, \
         host_match_on_header):

    try:
        endpoint = IngressInfo()
        endpoint.load_config_mode = load_config_mode
        endpoint.include_label_selectors = include_label_selectors
        endpoint.exclude_label_selectors = exclude_label_selectors
        endpoint.namespaces = namespaces
        endpoint.host_match_on_header = host_match_on_header
        if enable_cache:
            logging.info("init() caching enabled @ %s (TTL %d seconds)" % (cache_dir, cache_ttl_seconds))
            endpoint.cache = Cache(cache_dir)
            endpoint.cache_ttl_seconds = cache_ttl_seconds
        endpoint.init()
        endpoints.serverFromString(reactor, "tcp:" + str(listen_port)).listen(server.Site(endpoint))


        # start it up
        httpdthread = threading.Thread(target=reactor.run,args=(False,))
        httpdthread.daemon = True
        httpdthread.start()

        logging.info("init() listening on %d" % listen_port)

        try:
            while True:
                time.sleep(30)
        except KeyboardInterrupt:
            print("Exiting...")


    except Exception as e:
        logging.exception("init() unexpected error: " +  str(sys.exc_info()))
        sys.exit(1)


###########################
# Main program
##########################
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('-r', '--load-config-mode', dest='load_config_mode', default="local", \
        help="How the target k8s cluster config will be loaded, 'local' will leverage the current kubectl context from '~/.kube/config', while 'cluster' will talk direct to the cluster the process is executing in using the Pod's configured serviceAccount which needs read access to all objects of type Ingress. Default 'local'")
    parser.add_argument('-i', '--include-label-selectors', dest="include_label_selectors", default=None, \
        help="Optional comma delimited of Ingress label1=value,label2=value pairs that will be used to build the database of IngressInfo objects available to be fetched. If specified, ONLY Ingress objects having ALL specified labels will be retrieved. If not specified, ALL available Ingress objects will be retrieved from k8s. Excludes take precendence over includes.")
    parser.add_argument('-x', '--exclude-label-selectors', dest="exclude_label_selectors", default=None, \
        help="Optional comma delimited of Ingress label1=value,label2=value pairs that will be used to restrict the database of IngressInfo objects available to be fetched. If specified, ONLY Ingress objects NOT having ANY of the specified labels will be retrieved. If not specified, ALL available Ingress objects will be retrieved from k8s. Excludes take precendence over includes.")
    parser.add_argument('-n', '--namespace', dest="namespaces", default=None, \
        help="Optional comma delimited of Namespaces to scope Ingress fetch within")
    parser.add_argument('-m', '--host-match-on-header', dest="host_match_on_header", default=None, \
        help="Optional name of HTTP Request header who's value will be inspected for an Ingress host match, should the default path based /[host] not find a match.")
    parser.add_argument('-p', '--listen-port', dest='listen_port', \
        help="Port to listen on, default 8081", type=int, default=8081)
    parser.add_argument('-c', '--enable-cache', action='store_true', default=False, \
        help="Optional, enabling caching (uses https://github.com/grantjenks/python-diskcache)")
    parser.add_argument('-e', '--cache-ttl-seconds', dest="cache_ttl_seconds", type=int, default=30, \
        help="Optional, cache TTL in seconds")
    parser.add_argument('-d', '--cache-dir', dest="cache_dir", default="/opt/kubernetes-ingress-info/cache", \
        help="Optional, cache dir, default /opt/kubernetes-ingress-info/cache")
    parser.add_argument('-l', '--log-level', dest='log_level', default="DEBUG", \
        help="log level, default DEBUG ")
    parser.add_argument('-b', '--log-file', dest='log_file', default=None, \
        help="Path to log file, default None, STDOUT")

    args = parser.parse_args()

    dump_help = False

    if dump_help:
        parser.print_help()
        sys.exit(1)

    logging.basicConfig(level=logging.getLevelName(args.log_level),
                        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                        filename=args.log_file,filemode='w')
    logging.Formatter.converter = time.gmtime

    dictExcludeLabelSelectors = None
    if args.exclude_label_selectors is not None:
        dictExcludeLabelSelectors = {}
        for raw in args.exclude_label_selectors.split(","):
            dictExcludeLabelSelectors[raw.split("=")[0]] = raw.split("=")[1]

    listNamespaces = None
    if args.namespaces is not None:
        listNamespaces = args.namespaces.split(",")

    # invoke!
    init(args.load_config_mode, \
         args.listen_port,
         args.enable_cache,
         args.cache_dir,
         args.cache_ttl_seconds,
         args.include_label_selectors,
         dictExcludeLabelSelectors,
         listNamespaces,
         args.host_match_on_header)
